File: HeartForCharity/UI/heartforcharity_mobile/lib/model/responses/volunteer_job.dart
import 'package:heartforcharity_mobile/model/responses/skill.dart';

class VolunteerJob {
  final int volunteerJobId;
  final int organisationProfileId;
  final String organisationName;
  final int? categoryId;
  final String? categoryName;
  final String title;
  final String? description;
  final List<Skill> requiredSkills;
  final DateTime? startDate;
  final DateTime? endDate;
  final bool isRemote;
  final int positionsAvailable;
  final int positionsFilled;
  final int positionsRemaining;
  final String status;
  final int? addressId;
  final String? cityName;
  final DateTime createdAt;

  VolunteerJob({
    required this.volunteerJobId,
    required this.organisationProfileId,
    required this.organisationName,
    this.categoryId,
    this.categoryName,
    required this.title,
    this.description,
    this.requiredSkills = const [],
    this.startDate,
    this.endDate,
    required this.isRemote,
    required this.positionsAvailable,
    required this.positionsFilled,
    required this.positionsRemaining,
    required this.status,
    this.addressId,
    this.cityName,
    required this.createdAt,
  });

  factory VolunteerJob.fromJson(Map<String, dynamic> json) {
    final available = (json['positionsAvailable'] as num?)?.toInt() ?? 0;
    final filled = (json['positionsFilled'] as num?)?.toInt() ?? 0;
    return VolunteerJob(
      volunteerJobId: json['volunteerJobId'],
      organisationProfileId: json['organisationProfileId'],
      organisationName: json['organisationName'],
      categoryId: json['categoryId'],
      categoryName: json['categoryName'],
      title: json['title'],
      description: json['description'],
      requiredSkills: (json['requiredSkills'] as List<dynamic>?)
              ?.map((s) => Skill.fromJson(s as Map<String, dynamic>))
              .toList() ??
          [],
      startDate: json['startDate'] != null ? DateTime.parse(json['startDate']) : null,
      endDate: json['endDate'] != null ? DateTime.parse(json['endDate']) : null,
      isRemote: json['isRemote'] ?? false,
      positionsAvailable: available,
      positionsFilled: filled,
      positionsRemaining: (json['positionsRemaining'] as num?)?.toInt() ?? (available - filled),
      status: json['status']?.toString() ?? '',
      addressId: json['addressId'],
      cityName: json['cityName'],
      createdAt: DateTime.parse(json['createdAt']),
    );
  }
}

File: HeartForCharity/HeartForCharity.Services/Database/VolunteerJob.cs
using HeartForCharity.Model.Enums;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace HeartForCharity.Services.Database
{
    public class VolunteerJob
    {
        [Key]
        public int VolunteerJobId { get; set; }

        [Required]
        [ForeignKey(nameof(OrganisationProfile))]
        public int OrganisationProfileId { get; set; }

        [ForeignKey(nameof(Category))]
        public int? CategoryId { get; set; }

        [ForeignKey(nameof(Address))]
        public int? AddressId { get; set; }

        [Required]
        [MaxLength(200)]
        public string Title { get; set; } = null!;

        [MaxLength(4000)]
        public string? Description { get; set; }

        public DateTime? StartDate { get; set; }
        public DateTime? EndDate { get; set; }

        public bool IsRemote { get; set; } = false;

        public int PositionsAvailable { get; set; }
        public int PositionsFilled { get; set; } = 0;

        public VolunteerJobStatus Status { get; set; } = VolunteerJobStatus.Active;

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? DeletedAt { get; set; }

        // Navigacijska svojstva
        public virtual OrganisationProfile OrganisationProfile { get; set; } = null!;
        public virtual Category? Category { get; set; }
        public virtual Address? Address { get; set; }
        public virtual ICollection<VolunteerApplication> VolunteerApplications { get; set; } = new List<VolunteerApplication>();
        public virtual ICollection<VolunteerJobSkill> VolunteerJobSkills { get; set; } = new List<VolunteerJobSkill>();
    }
}